contributors/design-proposals/auth/node-tls-bootstrap-via-tpm.md

```diff
@@ -0,0 +1,280 @@
+cryptographic operations.
+
+Endorsement Key (EK) is a primary key for Endorsement Hierarchy baked into the
+TPM.
+EK Certificate is an x509 certificate created using EK and signed by an
+external Certificate Authority (CA).
+
+Attestation Identity Key (AIK) is a version of EK which is sign-only (as
+opposed to decrypt-only for default EK).
+AIK Certificate is EK Certificate with AIK as private key.
+
+### Current TLS cert bootstrapping
+
+[Node TLS
+bootstrapping](https://kubernetes.io/docs/admin/kubelet-tls-bootstrapping/) is
+a process of issuing client TLS certificates to a new k8s Node (kubelet) for
+authorization to API server.
+
+Current Kubelet TLS bootstrapping documentation suggests using a bootstrap
+bearer token and either using the [“nodeclient” SubjectAccessReview
+(SAR)](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/approver/sarapprove.go)
+approval flow or doing manual approvals. These present security issues:
+
+* Weak bootstrap credentials: Bootstrap bearer tokens are bad for all the
+  reasons bearer tokens are bad. They tend to be long lived (or static), and
+  not single use. When used in combination with “nodeclient” SAR approval
+  (discussed below), a compromised token allows an attacker join the cluster
+  from arbitrary machines as a new Node.
+* Weak approval flows:
+  * “nodeclient” SAR approval is bad because it allows any node in the cluster
+    to get credentials for any other node in the cluster.
+  * Manual approval is bad because it doesn’t scale and operators make errors.
+
+While other models of node TLS bootstrap are possible with the current
+implementation, they are much more difficult to setup, undocumented and thus
+not widely deployed.
+
+### TPM
+
+Private keys never leave the TPM boundaries in plain text.
+
+TPMs can also generate keys and organize them in hierarchies, such that
+parent-child relationship can be cryptographically proven.
+
+Hardware TPMs are usually low-power chips with very limited storage and poor
+performance. Software TPMs also have limited storage but can perform
+cryptographic operations much faster.
+
+## Overview
+
+In the proposed design, the kubelet can use a TPM as proof of identity to API
+server when creating TLS key pair used to authenticate to API server.
+
+At machine creation time, the TPM can be provisioned with AIK Certificate
+signed by a trusted CA. The certificate includes machine’s identifying info.
+Kubelet’s TLS keys will be attested by AIK to link them to machine identity.
+
+![Node bootstrap flow using TPM](node-tls-bootstrap-via-tpm.svg)
+
+After machine setup, node bootstrapping will go as follows:
+
+1. Kubelet generates an Elliptic Curve (EC) key pair and stores it on disk
+1. Kubelet triggers AIK creation in the TPM
+1. Kubelet requests that TPM certifies (signs) ECC public key with AIK
+1. Kubelet loads TPM’s AIK cert into memory
+1. Kubelet builds a Certificate Signing Request (CSR) using TLS key pair
```

Comment:
Just to clarify. "TLS key pair" is the same as the "ECC key pair" discussed earlier and no the AIK key pair, right?

Reply:
Yep, "TLS key pair" == "ECC key pair".